### ISSUE TITLE
Add support for Dockerfile.dockerignore files

### ISSUE THREAD

**Issue by nicks** [tilt-dev/tilt#3231]
From the tilt channel:
> Looks like tilt doesn’t support docker_build from Dockerfile.foo with related Dockerfile.foo.dockerignore. Can anybody confirm this ?

I haven't been able to find official specification of this feature, only a brief reference to it in release notes and in random github comments
https://github.com/moby/moby/issues/12886#issuecomment-480575928
and here
https://github.com/docker/engine/pull/215

It's supposed to be implemented on the buildkit side, though we need to dig in a bit about why it's not working out of the box

**Comment by nicks** [comment]
From [Slack #tilt](https://kubernetes.slack.com/archives/CESBL84MV/p1595805784300600) 

> buildkit has a feature where if you use a custom-named Dockerfile it will look for a .dockerignore that is prepended with the name of the Dockerfile before falling back on the normal .dockerignore (so Dockerfile.test will cause it to look for Dockerfile.test.dockerignore). When I run a build manually with DOCKER_BUILDKIT=1 the build is successful, but the same build done by tilt seems to always use the .dockerignore with no prepended name.

**Comment by nicks** [comment]
I looked into this a while back.

it's a bit tricky because there's no official specification of this behavior (AFAIK). On the docker side, it's implemented on the server-side (rather than client side), and the way it's implemented has weird edge cases. I wasn't sure how much of those edge cases to replicate of tilt. But maybe something simple is fine.

**Comment by salim-runsafe** [comment]
@nicks this is the last major hurdle keeping me from rolling out Tilt to our team so if there's anything I can do to help please let me know.

Here is the pull request where this feature was added to buildkit: https://github.com/moby/buildkit/pull/901

I agree that there doesn't seem to be an official spec, but it is mentioned there (and I can confirm) that

> Frontend will first check for <path/to/Dockerfile>.dockerignore and if it is found it will be used instead of the root .dockerignore

I don't really know go, but the limited code change seems to suggest that it wasn't too invasive of a change.

Of course since this is affecting me right now I would rather risk some edge cases to get the fix for what I need, but understand that you don't want to release something unstable.

**Comment by salim-runsafe** [comment]
I tried hacking at this a little and was able to get things to work for me with these changes:

**docker.go**
1. add `dbDockerfilePath string` to the list of arguments to `dockerignoresFromPathsAndContextFilters`
2. Added a little code after the existing `diFile := ...` on line 555

```        diFile := filepath.Join(path, ".dockerignore")
        customDiFile := dbDockerfilePath + ".dockerignore"
        _, err := os.Stat(customDiFile)
        if os.IsNotExist(err) {
        } else {
            diFile = customDiFile
        }```

3. Modified the call to `dockerignoresFromPathsAndContextFilters` within `dockerignoresForImage` to pass in `image.dbDockerfilePath` for the argument added in item 1, above

**docker_compose.go**
1. Modified the call to `dockerignoresFromPathsAndContextFilters` within `dcServiceToManifest` to pass in an empty string for the argument added in item 1, above

This was merely meant as a proof of concept and it's by no means production-ready, but if you think I'm heading in the right direction @nicks then I don't mind adding some tests around this and submitting an actual PR. I tested that it works as expected for my non-docker-compose use case and that adding and deleting the various .dockerignore files is handled gracefully by tilt as is.
